File: handal95/Programmers/등굣길.py
def solution(m, n, puddles):
    dp = [ [0] * (m+2) for row in range(n+2) ]
    dp[1][1] = 1
    
    for row in range(1, n + 1):
        for col in range(1, m + 1):
            if [col, row] in puddles:
                continue
                
            dp[row][col + 1] += dp[row][col] if [col + 1, row] not in puddles else 0
            dp[row + 1][col] += dp[row][col] if [col, row + 1] not in puddles else 0

    answer = dp[n][m] % 1000000007
    return answer

File: GyuZzang/프로그래머스/등굣길.cpp
#include <string>
#include <vector>

using namespace std;

int solution(int m, int n, vector<vector<int>> puddles) {
    vector<vector<int>> grid;
    grid.resize(n+1, vector<int> (m+1, 0));

    for(auto puddle : puddles){
        grid[puddle[1]-1][puddle[0]-1] = -1;
    }

    grid[0][0]=1;

    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            if(grid[i][j] >= 0) {
                if(grid[i][j+1]>=0) grid[i][j+1] = (grid[i][j+1] + grid[i][j]) %  1000000007;
                if(grid[i+1][j]>=0) grid[i+1][j] = (grid[i+1][j] + grid[i][j]) %  1000000007;
            }
        }
    }
    return grid[n-1][m-1];
}
